output/pixel_pusher.c

```diff
@@ -0,0 +1,272 @@
+    if (out_fd > 0) {
+        close(out_fd);
+    }
+    out_fd = socket(AF_INET, SOCK_DGRAM, 0);
+    if (out_fd < 0) {
+        PERROR("Error opening PixelPusher data socket");
+        return -1;
+    }
+
+    // Target address
+    // This is only called if pp_find_pusher finishes successfully, so
+    // pp_info is set
+    out_addr = calloc(1, sizeof *out_addr);
+    out_addr->sin_family = AF_INET;
+    out_addr->sin_addr.s_addr = pp_info->header.ip_addr;
+    out_addr->sin_port = htons(pp_info->info.my_port);
+
+    out_packet = calloc(1, 4 + 2*(1+3*240));
+    if (out_packet == NULL) MEMFAIL();
+
+    return 0;
+}
+
+int output_pp_init() {
+    if (pp_find_pusher() < 0) {
+        return -1;
+    }
+
+    if (pp_init_out() < 0) {
+        return -1;
+    }
+
+    return pp_add_grids();
+}
+
+void output_pp_term() {
+    INFO("Terminating PixelPusher");
+
+    if (pp_info) {
+        free(pp_info);
+        pp_info = NULL;
+    }
+
+    for (size_t i = 0; i < n_grid_devices; i++) {
+        struct output_device base = grid_devices[i].base;
+
+        free(base.pixels.xs);
+        free(base.pixels.ys);
+        free(base.pixels.colors);
+
+        if (base.prev != NULL)
+            base.prev->next = base.next;
+        if (base.next != NULL)
+            base.next->prev = base.prev;
+    }
+    free(grid_devices);
+    grid_devices = NULL;
+    n_grid_devices = 0;
+
+
+    if (out_fd) {
+        close(out_fd);
+        out_fd = -1;
+    }
+
+    if (out_addr) {
+        free(out_addr);
+        out_addr = NULL;
+    }
+
+    if (out_packet) {
+        free(out_packet);
+        out_packet = NULL;
+    }
+}
+
+int output_pp_do_frame() {
+    seq_num++;
+
+    ((uint32_t *) out_packet)[0] = seq_num;
+    size_t out_packet_idx = 4;
+
+    for (size_t i = 0; i < n_grid_devices; i++) {
+        struct pp_device * device = &grid_devices[i];
+
+        // Make sure it was successfully initialized
+        if (!device->base.active) continue;
+
+        // Copy the data to send into a buffer - sadly SDL_Color is rgba
+        // so we can't just send a header and it using sendmsg
+        out_packet[out_packet_idx++] = device->strip_num;
+
+        // Snake: The PixelPusher has linear strips arranged into a grid
+        // by going "back and forth".
+        for (int j = 0; j < device->height; j++) {
+            for (int k = 0; k < device->width; k++) {
+                int idx = (j % 2 ? device->width - 1 - k: k)*(device->height) + j;
+
+                SDL_Color color = device->base.pixels.colors[idx];
+                double alpha = color.a / 255.0;
+                out_packet[out_packet_idx++] = color.r * alpha;
+                out_packet[out_packet_idx++] = color.g * alpha;
+                out_packet[out_packet_idx++] = color.b * alpha;
+            }
+        }
+
+        // Send it: the PixelPusher can take 2 grids per packet
+        if ((i % 2) || (i == n_grid_devices - 1)) {
+            printf("Sending packet with i=%zu\n", i);
```

Comment:
Yeah this isn't really ever necessary; I just forgot to remove it before committing :-(